nestkernel/eprop_archiving_node.h

```diff
@@ -117,21 +199,99 @@ class EpropArchivingNodeRecurrent : public EpropArchivingNode< HistEntryEpropRec
+  //! Select the surrogate gradient function.
+  double ( EpropArchivingNodeRecurrent::*select_surrogate_gradient(
+    std::string surrogate_gradient_function ) )( double, double, double, double, double, double )
+  {
+    const std::map< std::string,
+      double ( EpropArchivingNodeRecurrent::* )( double, double, double, double, double, double ) >
+      surrogate_gradient_funcs = { { "piecewise_linear",
+                                     &EpropArchivingNodeRecurrent::compute_piecewise_linear_surrogate_gradient },
+        { "exponential", &EpropArchivingNodeRecurrent::compute_exponential_surrogate_gradient },
+        { "fast_sigmoid_derivative", &EpropArchivingNodeRecurrent::compute_fast_sigmoid_derivative_surrogate_gradient },
+        { "arctan", &EpropArchivingNodeRecurrent::compute_arctan_surrogate_gradient } };
+
+    auto found_entry_it = surrogate_gradient_funcs.find( surrogate_gradient_function );
+    if ( found_entry_it == surrogate_gradient_funcs.end() )
+    {
+      std::string error_message = "Surrogate gradient / pseudo-derivate function surrogate_gradient_function from [";
+      for ( const auto& surrogate_gradient_func : surrogate_gradient_funcs )
+      {
+        error_message += " \"" + surrogate_gradient_func.first + "\",";
+      }
+      error_message.pop_back();
+      error_message += " ] required.";
+
+      throw BadProperty( error_message );
+    }
+    else
+    {
+      return found_entry_it->second;
+    }
+  }
+
+  //! Compute the surrogate gradient with a piecewise linear function around the spike time (used, e.g., in Bellec et
+  //! al., 2020).
+  double compute_piecewise_linear_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with an exponentially decaying function around the spike time (used, e.g., in
+  //! Shrestha and Orchard, 2018).
+  double compute_exponential_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with a function corresponding to the derivative of a fast sigmoid around the spike
+  //! (used, e.g., in Zenke and Ganguli, 2018).
+  double compute_fast_sigmoid_derivative_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with an arctan function around the spike time (used, e.g., in Fang et al., 2021).
+  double compute_arctan_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Create an entry for the given time step at the end of the eprop history.
+  void emplace_new_eprop_history_entry( const long time_step );
```

Comment:
"emplace" has a specific meaning in modern C++. I would suggest to call it `store_...` or `register_...` or, since it is at the end, `append_...`

Reply:
Fixed in [5356e64](https://github.com/nest/nest-simulator/pull/3207/commits/5356e6478d2f0c2410723d8ce5b146036fa418a7)